Eliminate global usage of torch.set_default_dtype in test_autograd

aocsa opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56446 Eliminate global usage of torch.set_default_dtype in test_autograd**

Differential Revision: [D28000589](https://our.internmc.facebook.com/intern/diff/D28000589)

aocsa committed:
Eliminate global usage of torch.set_default_dtype in test_autograd

[ghstack-poisoned]

aocsa committed:
Update on "Eliminate global usage of torch.set_default_dtype in test_autograd"




[ghstack-poisoned]

aocsa committed:
Update on "Eliminate global usage of torch.set_default_dtype in test_autograd"




[ghstack-poisoned]

aocsa committed:
Update on "Eliminate global usage of torch.set_default_dtype in test_autograd"




[ghstack-poisoned]

mruberry commented:
@albanD would you like to take a look? This PR is part of our goal of getting rid of set_default_dtype() in the test suite; it's not so bad in test_autograd but has led to unexpected behavior in the past

aocsa committed:
Update on "Eliminate global usage of torch.set_default_dtype in test_autograd"




[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -774,13 +770,13 @@ def bw_hook(grad):
 
         self.assertEqual(counter[0], 1, msg='bw_hook not called')
-        self.assertEqual(x.grad, torch.ones(5, 5) * 2, atol=1e-5, rtol=0)
+        self.assertEqual(x.grad, torch.ones(5, 5, dtype=torch.float) * 2, atol=0.01, rtol=0)
```
This should be double and keeping the old tolerance I think.

aocsa replied:
I tried to maintain old tolerance and `dtype=torch.double` here, however when I removed global `torch.set_default_dtype(torch.double)` the `torch.nn.BatchNorm1d` class generates this error:  

```
torch/nn/functional.py", line 2261, in batch_norm
    return torch.batch_norm(
RuntimeError: expected scalar type Double but found Float
```
So I guess in the internal implementation there are some allocation with `dtype=torch.float`. 
I've adjusted the tolerance to a closest value `atol=1e-4, rtol=0`.

albanD replied:
Ho you need to add a `bn.double()` after creating it to move the Module to the right dtype.

albanD commented on test/test_autograd.py:
```diff
@@ -1020,44 +1016,45 @@ def reset_grad():
         torch.autograd.backward(fn(), gradient, inputs=[x])
         self.assertEqual(x.grad, x_grad_expected)
-        self.assertEqual(y.grad, torch.zeros(2, 2))
+        self.assertEqual(y.grad, torch.zeros(2, 2, dtype=torch.double))
```
nit: do we prefer here to pass `exact_dtype=False` here instead of creating double Tensors?
Same for a few below.

albanD reviewed (commented):
The code update looks good to me.
Let's see what the CI thinks :)

aocsa committed:
Update on "Eliminate global usage of torch.set_default_dtype in test_autograd"




[ghstack-poisoned]

aocsa committed:
Update on "Eliminate global usage of torch.set_default_dtype in test_autograd"




[ghstack-poisoned]

aocsa committed:
Update on "Eliminate global usage of torch.set_default_dtype in test_autograd"




[ghstack-poisoned]

aocsa committed:
Update on "Eliminate global usage of torch.set_default_dtype in test_autograd"




[ghstack-poisoned]

aocsa committed:
Update on "Eliminate global usage of torch.set_default_dtype in test_autograd"




[ghstack-poisoned]

aocsa committed:
Update on "Eliminate global usage of torch.set_default_dtype in test_autograd"




[ghstack-poisoned]

aocsa committed:
Update on "Eliminate global usage of torch.set_default_dtype in test_autograd"




[ghstack-poisoned]

albanD reviewed (approved):
Thanks for the update. CI seems happy with that so SGTM

pearu reviewed (changes requested):
I have a nit, otherwise LGTM.

aocsa commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5590,6 +5590,8 @@ def maybe_non_contig(tensor):
         # double check casting
         elif isinstance(arg, non_differentiable):
             if isinstance(arg.tensor, torch.Tensor):
+                if arg.tensor.dtype == torch.float:
+                    return maybe_non_contig(arg.tensor.to(dtype=dtype, device=device))
                 return maybe_non_contig(arg.tensor.to(device=device))
```
Yes there is a reason to transform to `torch.double`, as you can see `create_input` is used in `test_autograd/add_test` https://github.com/pytorch/pytorch/blob/70d9be0f42e569d6c74091bbed89ac2d9933ee65/test/test_autograd.py#L5478  to create in runtime new tests for inputs listed in `method_tests` https://github.com/pytorch/pytorch/blob/70d9be0f42e569d6c74091bbed89ac2d9933ee65/torch/testing/_internal/common_methods_invocations.py#L5348
Some of these inputs by default create tensors with `torch.float` dtype. So as I removed  `torch.set_default_dtype in test_autograd`,  conversion to dtype double is needed in the other case some tests will fail.

pearu replied:
I can see the need to transform to `torch.double`. However, if
```
return maybe_non_contig(arg.tensor.to(dtype=dtype, device=device))
```
is used, doesn't that guarantee that already? To be explicit, this is my suggestion:
```
        ...
        elif isinstance(arg, non_differentiable):
            if isinstance(arg.tensor, torch.Tensor):
                return maybe_non_contig(arg.tensor.to(dtype=dtype, device=device))
        ...
```
Notice that `dtype` is here either `torch.double` or `torch.cdouble`. Hence, if `arg.tensor.dtype` is already `torch.double`, the `to` method is no-op with respect to `dtype` argument.

aocsa replied:
I get what u mean, however I tried this before, and `test_jit/test_nn_multi_margin_loss` fails. 
```
FAILED test/test_jit.py::TestJitGeneratedFunctional::test_nn_multi_margin_loss - RuntimeError: expected scalar type Long but found Double
```
I think that two special case should be there, case float->double and possible case cfloat -> cdouble I will add another elif for this case.    cc @pearu

pearu replied:
The error message hints that something calling `create_input` does not specify `dtype` while it should set `dtype=torch.long`. It seems that test_jit.py has been in flux lately and finding this something is convolved (test_jit.py in the master is different from here), and the extent of this problem is unknown as well. So this issue might be OT for this PR..

aocsa committed:
Update on "Eliminate global usage of torch.set_default_dtype in test_autograd"




[ghstack-poisoned]

aocsa committed:
Update on "Eliminate global usage of torch.set_default_dtype in test_autograd"




[ghstack-poisoned]

pearu reviewed (approved):
LGTM

mruberry commented:
Sorry @aocsa, this PR just needs a rebase

aocsa committed:
Update on "Eliminate global usage of torch.set_default_dtype in test_autograd"


Differential Revision: [D28000589](https://our.internmc.facebook.com/intern/diff/D28000589)

[ghstack-poisoned]

mruberry commented:
Sorry @aocsa, another merge conflict (although Github doesn't seem to think so). Would you try rebasing this once more and I'll sneak it in ASAP

aocsa commented:
> Sorry @aocsa, another merge conflict (although Github doesn't seem to think so). Would you try rebasing this once more and I'll sneak it in ASAP

Don't worry, I see what is happening I will rebase. Once is done I will tag u.

mruberry commented:
> > Sorry @aocsa, another merge conflict (although Github doesn't seem to think so). Would you try rebasing this once more and I'll sneak it in ASAP
> 
> Don't worry, I see what is happening I will rebase. Once is done I will tag u.

Thank you!

aocsa committed:
Update on "Eliminate global usage of torch.set_default_dtype in test_autograd"


Differential Revision: [D28000589](https://our.internmc.facebook.com/intern/diff/D28000589)

[ghstack-poisoned]

aocsa commented:
> > > Sorry @aocsa, another merge conflict (although Github doesn't seem to think so). Would you try rebasing this once more and I'll sneak it in ASAP
> > 
> > 
> > Don't worry, I see what is happening I will rebase. Once is done I will tag u.
> 
> Thank you!

It should be done. I double-check the affected test_autograd.py test. Let see what CI says  cc @mruberry.

aocsa committed:
Update on "Eliminate global usage of torch.set_default_dtype in test_autograd"


Differential Revision: [D28000589](https://our.internmc.facebook.com/intern/diff/D28000589)

[ghstack-poisoned]

mruberry commented:
I've asked @albanD to take a look at another internal failure, sorry for the delay, @aocsa

aocsa committed:
Update on "Eliminate global usage of torch.set_default_dtype in test_autograd"


Differential Revision: [D28000589](https://our.internmc.facebook.com/intern/diff/D28000589)

[ghstack-poisoned]

aocsa committed:
Update on "Eliminate global usage of torch.set_default_dtype in test_autograd"


Differential Revision: [D28000589](https://our.internmc.facebook.com/intern/diff/D28000589)

[ghstack-poisoned]
